ports/samd/samd_spiflash.c

```diff
@@ -0,0 +1,311 @@
+ *
+ * Copyright (c) 2023 Robert Hammelrath
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ * Based on spiflash.py by Christopher Arndt and the EEPROM library of Peter Hinch
+ *
+ * https://github.com/SpotlightKid/micropython-stm-lib/tree/master/spiflash
+ * https://github.com/peterhinch/micropython_eeprom.git
+ */
+
+#include <stdint.h>
+#include "py/obj.h"
+#include "py/runtime.h"
+#include "py/mphal.h"
+#include "extmod/machine_spi.h"
+#include "extmod/virtpin.h"
+#include "extmod/vfs.h"
+#include "modmachine.h"
+
+#if MICROPY_HW_SPIFLASH
+
+#define _READ_INDEX (0)
+#define _PROGRAM_PAGE_INDEX (1)
+#define _SECTOR_ERASE_INDEX (2)
+
+const uint8_t _CMDS3BA[] = {0x03, 0x02, 0x20};  // CMD_READ CMD_PROGRAM_PAGE CMD_ERASE_4K
+const uint8_t _CMDS4BA[] = {0x13, 0x12, 0x21};  // CMD_READ CMD_PROGRAM_PAGE CMD_ERASE_4K
+
+#define CMD_JEDEC_ID (0x9F)
+#define CMD_READ_STATUS (0x05)
+#define CMD_WRITE_ENABLE (0x06)
+#define CMD_READ_UID (0x4B)
+#define CMD_READ_SFDP (0x5A)
+#define PAGE_SIZE (256)
+#define SECTOR_SIZE (4096)
+
+typedef struct _spiflash_obj_t {
+    mp_obj_base_t base;
+    mp_obj_base_t *spi;
+    mp_obj_base_t *cs;
+    bool addr4b;
+    uint16_t pagesize;
+    uint16_t sectorsize;
+    const uint8_t *commands;
+    uint32_t size;
+} spiflash_obj_t;
+
+extern const mp_obj_type_t spiflash_type;
+
+static void spi_transfer(mp_obj_base_t *spi, size_t len, const uint8_t *src, uint8_t *dest) {
+    mp_machine_spi_p_t *spi_p = (mp_machine_spi_p_t *)MP_OBJ_TYPE_GET_SLOT(spi->type, protocol);
+    spi_p->transfer(spi, len, src, dest);
+}
+
+static void wait(spiflash_obj_t *self) {
+    uint8_t msg[2];
+    uint32_t timeout = 100000;
+
+    // each loop takes at least about 5µs @ 120Mhz. So a timeout of
+    // 100000 wait 500ms max. at 120Mhz. Sector erase lasts about
+    // 100ms worst case, page write is < 1ms.
+    do {
+        msg[0] = CMD_READ_STATUS;
+        mp_virtual_pin_write(self->cs, 0);
```

Comment:
I'd suggest using `mp_hal_pin_write(self->cs, 0)`.  It should be smaller (and slightly more efficient) and I don't think you need the generality of the virtual pin API (it's only needed if you want to pass `machine.Signal` objects in).

Reply:
The intention was to make it port independent. mp_hal_pin_write() is slightly different on the various ports.

Reply:
But `mp_hal_pin_write()` should still be portable and work the same way on all ports.  Other drivers (eg dht, soft I2C, soft SPI) use this.

Reply:
Yes, but the type of the first argument differs between ports. it's mp_hal_pin_obj_t, but that is an mp_obt_t kind on some ports, on others it's an int. Nevertheless, it's easy to change it.

Reply:
Changed